src/Storages/ObjectStorage/StorageObjectStorage.h

```diff
@@ -254,10 +253,11 @@ class StorageObjectStorage::Configuration
         throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Method createArgsWithAccessData is not supported by storage {}", getEngineName());
     }
 
-protected:
     virtual void fromNamedCollection(const NamedCollection & collection, ContextPtr context) = 0;
     virtual void fromAST(ASTs & args, ContextPtr context, bool with_structure) = 0;
 
+    virtual ObjectStorageType extractDynamicStorageType(ASTs & /* args */, ContextPtr /* context */, ASTPtr * /* type_arg */ = nullptr) const
```

Comment:
Is there a static storage type as well?

Reply:
No, it returns `S3`, `Azure`, `HDFS`, `Local` for different `storage_type` parameter if it exists.